Signal/src/AppDelegate.m

```diff
@@ -564,33 +566,31 @@ - (void)application:(UIApplication *)application
                completionHandler:(void (^)(BOOL succeeded))completionHandler {
     OWSAssertIsOnMainThread();
 
-    if (!AppReadiness.isAppReady) {
-        DDLogWarn(@"%@ Ignoring performActionForShortcutItem: app not ready.", self.logTag);
-        // TODO: Consider using [AppReadiness runNowOrWhenAppIsReady:].
-        completionHandler(NO);
-    }
+    [AppReadiness runNowOrWhenAppIsReady:^{
+        if (![TSAccountManager isRegistered]) {
+            UIAlertController *controller =
+                [UIAlertController alertControllerWithTitle:NSLocalizedString(@"REGISTER_CONTACTS_WELCOME", nil)
+                                                    message:NSLocalizedString(@"REGISTRATION_RESTRICTED_MESSAGE", nil)
+                                             preferredStyle:UIAlertControllerStyleAlert];
+
+            [controller addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil)
+                                                           style:UIAlertActionStyleDefault
+                                                         handler:^(UIAlertAction *_Nonnull action){
+
+                                                         }]];
+            UIViewController *fromViewController = [[UIApplication sharedApplication] frontmostViewController];
+            [fromViewController presentViewController:controller
+                                             animated:YES
+                                           completion:^{
+                                               completionHandler(NO);
+                                           }];
+            return;
+        }
 
-    if ([TSAccountManager isRegistered]) {
         [SignalApp.sharedApp.homeViewController showNewConversationView];
-        completionHandler(YES);
-    } else {
-        UIAlertController *controller =
-            [UIAlertController alertControllerWithTitle:NSLocalizedString(@"REGISTER_CONTACTS_WELCOME", nil)
-                                                message:NSLocalizedString(@"REGISTRATION_RESTRICTED_MESSAGE", nil)
-                                         preferredStyle:UIAlertControllerStyleAlert];
-
-        [controller addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil)
-                                                       style:UIAlertActionStyleDefault
-                                                     handler:^(UIAlertAction *_Nonnull action){
-
-                                                     }]];
-        UIViewController *fromViewController = [[UIApplication sharedApplication] frontmostViewController];
-        [fromViewController presentViewController:controller
-                                         animated:YES
-                                       completion:^{
-                                           completionHandler(NO);
-                                       }];
-    }
+    }];
+
+    completionHandler(YES);
```

Comment:
It seems like now we could call the completionHandler twice - always here, and sometimes again after presenting the alertController. Why this change in control flow?

Reply:
Good catch, this belongs inside the block above.